### PR TITLE
`ScopeOwned` in `sherlock.h`, 50% done.

### PR DESCRIPTION
**Pull request by dkorolev**
Hi Max,

PTAL after https://github.com/C5T/Current/pull/462. No need to merge, I'm happy to make this change atop yours later on.

Thanks,
Dima

### CODE REVIEW

File: Blocks/Persistence/memory.h [hunk 1]
```diff
@@ -74,26 +74,50 @@ class MemoryPersister {
...
+          CURRENT_THROW(PersistenceMemoryBlockNoLongerAvailable());
+        }
+        std::lock_guard<std::mutex> lock(container_->mutex);
+        return Entry(i_, container_->entries[i_]);
+      }
+      void operator++() {
+        if (!valid_) {
+          CURRENT_THROW(PersistenceMemoryBlockNoLongerAvailable());
+        }
+
```
Comment (mzhurovich) [inline Blocks/Persistence/memory.h]: Extra empty line.

File: Sherlock/pubsub.h [hunk 2]
```diff
@@ -205,6 +222,9 @@ class PubSubHTTPEndpointImpl {
   // TODO(dkorolev): This is a long shot, but looks right: For type-filtered HTTP subscriptions,
   // whether we should terminate or no depends on `nowait`.
   ss::EntryResponse EntryResponseIfNoMorePassTypeFilter() const {
+    if (time_to_terminate_) {
+      return ss::EntryResponse::Done;
+    }
     return no_wait_ ? ss::EntryResponse::Done : ss::EntryResponse::More;
```
Comment (mzhurovich) [inline Sherlock/pubsub.h]: `return (time_to_terminate_ || no_wait_) ? ss::EntryResponse::Done : ss::EntryResponse::More;`?

File: Sherlock/sherlock.h [hunk 3]
```diff
@@ -62,57 +61,17 @@ SOFTWARE.
 //
 // Sherlock streams can be published into and subscribed to.
 //
-// Publishing is done via `my_stream.Publish(ENTRY{...});`. It is the caller's responsibility
+// Publishing is done via `my_stream.Publish(ENTRY{...});`.
 // to ensure publishing is done from one thread only (Sherlock itself offers no locking).
```
Comment (mzhurovich) [inline Sherlock/sherlock.h]: I suppose this line should also be deleted.

File: Sherlock/sherlock.h [hunk 4]
```diff
@@ -139,111 +98,73 @@ template <typename ENTRY, template <typename> class PERSISTENCE_LAYER = DEFAULT_
...
    public:
-    explicit StreamPublisher(std::shared_ptr<StreamData> data) : data_(data) {}
+    explicit StreamPublisher(ScopeOwned<stream_data_t>& data) : data_(data, []() {}) {}
 
     template <current::locks::MutexLockStatus MLS>
-    idxts_t DoPublish(const ENTRY& entry, const std::chrono::microseconds us = current::time::Now()) {
-      current::locks::SmartMutexLockGuard<MLS> lock(data_->mutex);
+    idxts_t DoPublish(const entry_t& entry, const std::chrono::microseconds us = current::time::Now()) {
+      if (!data_) {
+        CURRENT_THROW(StreamInGracefulShutdownMode());
```
Comment (mzhurovich) [inline Sherlock/sherlock.h]: Btw, in Storage we have `StorageInGracefulShutdownException`. Let's define some common style for these exceptions :)

File: Sherlock/pubsub.h (hunk 2 above)
Comment (sompylasar) [inline Sherlock/pubsub.h]: `return ((time_to_terminate_ || no_wait_) ? ss::EntryResponse::Done : ss::EntryResponse::More);`?
```
return (
  (time_to_terminate_ || no_wait_)
    ? ss::EntryResponse::Done
    : ss::EntryResponse::More
);
```
?

File: Sherlock/pubsub.h [hunk 5]
```diff
@@ -114,11 +118,21 @@ SOFTWARE.
...
+
+  PubSubHTTPEndpointImpl(const std::string& subscription_id, ScopeOwned<stream_data_t>& data, Request r)
+      : data_(data, [this]() { time_to_terminate_ = true; }),
+        http_request_(std::move(r)),
+        http_response_(http_request_.SendChunkedResponse(
+            HTTPResponseCode.OK,
+            current::net::constants::kDefaultJSONContentType,
+            current::net::http::Headers({
+                {kSherlockHeaderCurrentSubscriptionId, subscription_id},
+                {kSherlockHeaderCurrentStreamSize, current::ToString(data_->persistence.Size())},
```
Comment (mzhurovich) [inline Sherlock/pubsub.h]: What is the use case for this header? Stream size grows and we set header value to the size at the moment when HTTP subscription was started.

File: Sherlock/test.cc [hunk 6]
```diff
@@ -497,8 +466,11 @@ TEST(Sherlock, SubscribeToStreamViaHTTP) {
     EXPECT_EQ("4", result.headers.Get("X-Current-Stream-Size"));
   }
 
+  // DIMA
+  return;
```
Comment (mzhurovich) [inline Sherlock/test.cc]: Remove?

File: Sherlock/test.cc [hunk 7]
```diff
@@ -497,8 +466,11 @@ TEST(Sherlock, SubscribeToStreamViaHTTP) {
...
   // Test `n`.
   EXPECT_EQ(s[0], HTTP(GET(base_url + "?n=1")).body);
+  return;
```
Comment (mzhurovich) [inline Sherlock/test.cc]: Remove?

File: Blocks/Persistence/memory.h (hunk 1 above)
Comment (dkorolev) [inline Blocks/Persistence/memory.h]: :+1: 954d5e8

File: Sherlock/pubsub.h (hunk 2 above)
Comment (dkorolev) [inline Sherlock/pubsub.h]: :+1: 954d5e8

File: Sherlock/sherlock.h (hunk 3 above)
Comment (dkorolev) [inline Sherlock/sherlock.h]: :+1: 954d5e8

File: Sherlock/pubsub.h (hunk 2 above)
Comment (dkorolev) [inline Sherlock/pubsub.h]: (@sompylasar -- I did it Max's way :-) )

File: Sherlock/sherlock.h (hunk 4 above)
Comment (dkorolev) [inline Sherlock/sherlock.h]: :+1: 954d5e8

File: Sherlock/pubsub.h (hunk 5 above)
Comment (dkorolev) [inline Sherlock/pubsub.h]: Discussed to leave it as is (and add some `X-Current-Most-Recent-Timestamp` later).

File: Sherlock/test.cc (hunk 6 above)
Comment (dkorolev) [inline Sherlock/test.cc]: :+1: 954d5e8

File: Sherlock/test.cc (hunk 7 above)
Comment (dkorolev) [inline Sherlock/test.cc]: :+1: 954d5e8

**Comment by dkorolev** [comment]
Thanks for the review && PTAL.

File: Sherlock/test.cc [hunk 8]
```diff
@@ -610,6 +579,71 @@ TEST(Sherlock, SubscribeToStreamViaHTTP) {
...
+    const auto result =
+        HTTP(ChunkedGET(base_url,
+                        [&subscription_id](const std::string& header, const std::string& value) {
+                          if (header == "X-Current-Stream-Subscription-Id") {
+                            subscription_id = value;
+                          }
+                        },
+                        [&chunks_count, &chunks](const std::string& chunk_body) {
+                          ++chunks_count;
+                          chunks.push_back(chunk_body);
```
Comment (mzhurovich) [inline Sherlock/test.cc]: Looks like `chunks` content is not used anywhere in this test.

File: Sherlock/pubsub.h (hunk 2 above)
Comment (sompylasar) [inline Sherlock/pubsub.h]: Huh okay. The point of my style is the increased readability of return+ternary and assign+ternary (emphasize that the returned/assigned value is the whole parenthesized expression which is ternary inside).

Comment (dkorolev) [inline Sherlock/pubsub.h]: Even if I wanted to, our `clang-format` would fix it right away:

```
dima@dima-i9: ~/github/dkorolev/Current/Sherlock(fix_sherlock) $ cat >1.h
return (
  (time_to_terminate_ || no_wait_)
    ? ss::EntryResponse::Done
    : ss::EntryResponse::More
);
dima@dima-i9: ~/github/dkorolev/Current/Sherlock(fix_sherlock) $ make indent 
/home/dima/github/dkorolev/Current/scripts/indent.sh
dima@dima-i9: ~/github/dkorolev/Current/Sherlock(fix_sherlock) $ cat 1.h 
return ((time_to_terminate_ || no_wait_) ? ss::EntryResponse::Done : ss::EntryResponse::More);
```

File: Sherlock/test.cc (hunk 8 above)
Comment (dkorolev) [inline Sherlock/test.cc]: :+1: cffd957

File: Sherlock/pubsub.h (hunk 2 above)
Comment (sompylasar) [inline Sherlock/pubsub.h]: At least it preserved the outer parens :)

**Pull request merged**